src/components/Menu.tsx

```diff
@@ -0,0 +1,351 @@
+    /**
+     * The children of the sub-menu specify the structure of the options in the sub-menu.
+     *
+     * See the component documentation below for details.
+     */
+    children: React.ReactNode
+}
+
+/**
+ * This component can be rendered alongside other `MenuItem` inside a `MenuList` in order to have
+ * a sub-menu.
+ *
+ * The options of the sub-menu must be provided inside another `MenuList` explicitly given as
+ * `children`:
+ *
+ * ```jsx
+ * <MenuItem label="Edit profile" />
+ * <SubMenu label="More options">
+ *   <MenuList>
+ *     <MenuItem label="Preferences" />
+ *     <MenuItem label="Sign out" />
+ *   </MenuList>
+ * </SubMenu>
+ * ```
+ *
+ * The `MenuList` needs to be provided to give more flexibility to customize it (e.g. you can pass
+ * a `className` and other props to it).
+ */
+const SubMenu = React.forwardRef<HTMLButtonElement, SubMenuProps>(function SubMenu(
+    { label, icon, children, ...props },
+    ref,
+) {
+    const { handleItemSelect: parentMenuItemSelect, ...state } = React.useContext(MenuContext)
+    const { hide: parentMenuHide } = state
+
+    const handleSubItemSelect = React.useCallback(
+        function handleSubItemSelect(value) {
+            parentMenuItemSelect(value)
+            parentMenuHide()
+        },
+        [parentMenuHide, parentMenuItemSelect],
+    )
+
+    //
+    // This may seem controversial but it seems to work pretty well.
+    //
+    // We're dynamically building a component here because reakit's way to compose menus is by
+    // passing a component in the `as` prop to its `MenuItem` as seen below. That seems convenient,
+    // but on the other hand it does not allow us to own the rendering of the menu items that open
+    // a sub-menu. We do not want these to be an arbitrary `MenuButton` but we want to control what
+    // we render inside.
+    //
+    // Our API to create sub-menus is leaner, allowing us to compose it all inline while keeping
+    // control of the content of menu items not being arbitrary but having a certain structure.
+    //
+    const SubMenuList = React.useCallback(
+        React.forwardRef<HTMLButtonElement, Reakit.MenuButtonProps>(function SubMenuList(
+            innerProps,
+            innerRef,
+        ) {
+            return (
+                <Menu onItemSelect={handleSubItemSelect}>
+                    <MenuButton {...innerProps} ref={innerRef}>
+                        {icon ? <span className="reactist_menuitem__icon">{icon}</span> : null}
+                        <span className="reactist_menuitem__label">{label}</span>
+                        <svg viewBox="0 0 43.3 50">
+                            <polygon points="43.3 25 0 0 0 50 43.3 25"></polygon>
+                        </svg>
+                    </MenuButton>
+                    {children}
+                </Menu>
+            )
+        }),
+        [children, label, icon, handleSubItemSelect],
+    )
```

Comment:
> it does not allow us to own the rendering of the menu items that open a sub-menu

I'm not sure if I understand what you mean here. But you can also use render props so as to avoid creating a new component in the render phase:

```jsx
<Reakit.MenuItem {...state} {...props} ref={ref}>
  {(itemHTMLProps) => ( // itemHTMLProps already contains ref
    <Menu>
      <MenuButton {...itemHTMLProps}>...</MenuButton>
      {children}
    </Menu>
  )}
</Reakit.MenuItem>
```

Something I think may cause bugs in your code is that `children`, `label`, `icon` and `handleSubItemSelect` should be memoized. Otherwise it's going to mount a new `SubMenuList` component on every render, losing internal state.

For example, in the example usage, `children` will be always different since `React.createElement()` always returns a different object.

```jsx
<SubMenu label="More options">
  <MenuList>
    <MenuItem label="Preferences" />
    <MenuItem label="Sign out" />
  </MenuList>
</SubMenu>
```

Reply:
Alternatively, you can also create `SubMenuList` in the module scope as usual and pass props using `Reakit.Menuitem`.

```jsx
<Reakit.MenuItem
  {...props}
  {...state}
  as={SubMenuList}
  ref={ref}
  label={label}
  icon={icon}
  handleSubItemSelect={handleSubItemSelect}
>
  {children}
</Reakit.MenuItem>
```

Those additional props will be passed to `SubMenuList` as they aren't consumed by `Reakit.MenuItem`.

Reply:
Another simpler option that may work:
```jsx
// inside SubMenu component
return (
  <Menu onItemSelect={handleSubItemSelect}>
    <Reakit.MenuItem
      {...state}
      {...props}
      ref={ref}
      as={MenuButton}
      label={label}
      icon={icon}
    />
    {children}
  </Menu>
);
```

Reply:
Yup, this was already not something I was convinced with. Additionally I may ditch the restrictive way of providing content to menu items and let it be a free form children prop.

Anyway, thanks a lot for your insights 🎉. I’ll take a look today and combined with the above it will probably not be needed.

Reply:
I went with the first option above, but alongside allowing the user to provide the content of the menu button (because I also had decided yesterday after pushing this that it would be best to leave it free and let users of the lib to add the icon or any other structure inside it themselves).